tla/Traceccfraft.tla

```diff
@@ -115,122 +115,61 @@ OneMoreMessage(msg) ==
     /\ ToConfigurations(<<TraceLog[1].msg.new_configuration>>) = configurations[TraceLog[1].msg.state.node_id]
 
-\* The following sub-actions all leave the variables unchanged, and a single, generic sub-action
- \* would be sufficient.  However, the sub-actions are useful for debugging, as they make sure
- \* the log event's identifier shows up in TLC counterexamples.
-IsEvent(e) ==
-    /\ TLCGet("level")' \in 1..Len(TraceLog)
-    /\ TraceLog[TLCGet("level")'].msg.function = e
-
-become_follower ==
-    /\ IsEvent("become_follower")
-    /\ UNCHANGED vars
-
-send_request_vote_response ==
-    /\IsEvent("send_request_vote_response")
-    /\ UNCHANGED vars
-
-send_append_entries_response ==
-    /\ IsEvent("send_append_entries_response")
-    /\ UNCHANGED vars
-
-commit ==
-    /\ IsEvent("commit")
-    /\ UNCHANGED vars
-
-add_configuration ==
-    /\ IsEvent("add_configuration")
-    /\ UNCHANGED vars
-    
-execute_append_entries_sync ==
-    /\ IsEvent("execute_append_entries_sync")
-    /\ UNCHANGED vars
-    
-TraceRcvUpdateTermReqVote ==
-    RcvUpdateTerm \cdot RcvRequestVoteRequest
-
-TraceRcvUpdateTermRcvRequestVoteResponse ==
-    RcvUpdateTerm \cdot RcvRequestVoteResponse
-
-TraceRcvUpdateTermReqAppendEntries ==
-    RcvUpdateTerm \cdot RcvAppendEntriesRequest
-
-TraceRcvUpdateTermRcvAppendEntriesResponse ==
-    RcvUpdateTerm \cdot RcvAppendEntriesResponse
-
-TraceConflictAppendEntriesRequestNoConflictAppendEntriesRequest ==
-    \E m \in Messages : 
-        /\ m.type = AppendEntriesRequest
-        /\ LET i == m.dest
-               j == m.source
-               logOk == \/ m.prevLogIndex = 0
-                        \/ /\ m.prevLogIndex > 0
-                           /\ m.prevLogIndex <= Len(log[i])
-                           /\ m.prevLogTerm = log[i][m.prevLogIndex].term
-           IN 
-            /\ m.term = currentTerm[m.dest]
-            /\ state[m.dest] \in {Follower, Pending}
-            /\ logOk
-            /\ ConflictAppendEntriesRequest(m.dest, m.prevLogIndex + 1, m) 
-                \cdot NoConflictAppendEntriesRequest(m.dest, m.source, m)  
-            /\ UNCHANGED <<candidateVars, leaderVars>>
-
-TraceNext ==
-    \/ Next
-    
-    \/ become_follower
-    \/ send_request_vote_response
-    \/ send_append_entries_response
-    \/ commit
-    \/ add_configuration
-    \/ execute_append_entries_sync
-
-    \/ TraceRcvUpdateTermReqVote
-    \/ TraceRcvUpdateTermRcvRequestVoteResponse
-    \/ TraceRcvUpdateTermReqAppendEntries
-    \/ TraceRcvUpdateTermRcvAppendEntriesResponse
-    \/ TraceConflictAppendEntriesRequestNoConflictAppendEntriesRequest
-
-TraceSpec ==
-    TraceInit /\ [][TraceNext]_vars
+TruncateLog(i) ==
+    \* \* raft.h: recv_append_entries > become_aware_of_new_term > become_follower > rollback
+    \* TODO This action should be moved to the ccfraft module!
+    /\ log[i] # <<>> 
+    /\ commitIndex[i] < Len(log[i])
+    /\ log' = [log EXCEPT ![i] = SubSeq(log[i], 1, commitIndex[i])]
```

Comment:
As per discussion at https://github.com/microsoft/CCF/pull/5233#discussion_r1185834287, this will have to change later.